File: app/javascript/channels/chat_room_channel.js
// Action Cableのconsumerをインポート
// これによりWebSocket接続の管理が可能になる
import consumer from "./consumer"

// Turboによるページ読み込み完了時のイベントリスナーを設定
document.addEventListener('turbo:load', () => {
  // data-chat-room-id属性を持つ要素を探す
  // この要素にはチャットルームの識別子が含まれている
  const chatRoomElement = document.querySelector('[data-chat-room-id]')
  
  if (chatRoomElement) {
    // チャットルームのIDを取得
    const chatRoomId = chatRoomElement.dataset.chatRoomId

    // ChatRoomChannelへの購読を作成
    // chat_room_idパラメータを指定してチャンネルを特定
    consumer.subscriptions.create({ channel: "ChatRoomChannel", chat_room_id: chatRoomId }, {
      // WebSocket接続が確立された時の処理
      connected() {
        console.log("Connected to chat room channel")
      },

      // WebSocket接続が切断された時の処理
      disconnected() {
        console.log("Disconnected from chat room channel")
      },

      // サーバーからデータを受信した時の処理
      received(data) {
        if (data.type === 'message_read') {
          // メッセージが既読になった場合の処理
          // 該当するメッセージ要素を探して既読表示に更新
          const messageElement = document.querySelector(`[data-message-id="${data.message_id}"]`)
          if (messageElement) {
            messageElement.classList.add('message--read')
          }
        } else if (data.type === 'new_message') {
          // 新しいメッセージを受信した場合の処理
          // メッセージコンテナに新規メッセージを追加
          const messagesContainer = document.getElementById('messages')
          if (messagesContainer) {
            messagesContainer.insertAdjacentHTML('beforeend', data.message)
          }
        }
      },

      // メッセージを既読にするメソッド
      // サーバーサイドのmark_as_readアクションを呼び出す
      markAsRead(messageId) {
        this.perform('mark_as_read', { message_id: messageId })
      }
    })
  }
})

File: app/channels/chat_room_channel.rb
# チャットルーム機能のためのActionCableチャンネルクラス
# WebSocketを使用してリアルタイムな双方向通信を実現する
class ChatRoomChannel < ApplicationCable::Channel
  # クライアントがチャンネルを購読した時に呼ばれるメソッド
  # @param [Hash] params クライアントから送られてきたパラメータ
  # @option params [Integer] :chat_room_id 購読するチャットルームのID
  def subscribed
    # パラメータで指定されたIDのチャットルームを取得
    @chat_room = ChatRoom.find(params[:chat_room_id])
    # 指定されたチャットルームの通信ストリームを開始
    # このチャットルームに関連するメッセージをブロードキャストできるようになる
    stream_for @chat_room
  end

  # クライアントがチャンネルの購読を解除した時に呼ばれるメソッド
  # チャットルームから退出する際などに実行される
  def unsubscribed
    # 全ての通信ストリームを停止
    # メモリリークを防ぎ、不要な通信を避けるため
    stop_all_streams
  end

  # メッセージを既読状態にするメソッド
  # クライアントからメッセージIDを受け取り、該当メッセージを既読にする
  # @param [Hash] data クライアントから送られてきたデータ
  # @option data [Integer] 'message_id' 既読にするメッセージのID
  def mark_as_read(data)
    # 指定されたIDのメッセージを取得
    message = Message.find(data["message_id"])

    # メッセージが現在のチャットルームに属していて、
    # かつメッセージの送信者が現在のユーザーでない場合のみ処理を実行
    if message.chat_room == @chat_room && message.user != current_user
      # メッセージを既読状態にマーク
      message.mark_as_read!

      # チャットルーム内の全クライアントに既読状態の変更を通知
      # type: 'message_read' - クライアント側で既読表示を更新するために使用
      # message_id: - 既読になったメッセージを特定するためのID
      broadcast_to @chat_room, {
        type: "message_read",
        message_id: message.id
      }
    end
  end
end
